observer/baseNodeProvider.go

```diff
@@ -28,26 +29,173 @@ func (bop *baseNodeProvider) initNodesMaps(nodes []*data.NodeData) error {
+	bnp.mutNodes.RLock()
+	defer bnp.mutNodes.RUnlock()
+
+	nodesSlice := make([]*data.NodeData, 0)
+	for _, node := range bnp.outOfSyncNodes {
+		nodesSlice = append(nodesSlice, node)
+	}
+	for _, node := range bnp.allNodes {
+		nodesSlice = append(nodesSlice, node)
+	}
+
+	return nodesSlice
+}
+
+// UpdateNodesBasedOnSyncState will handle the nodes lists, by removing out of sync observers or by adding back observers
+// that were previously removed because they were out of sync.
+func (bnp *baseNodeProvider) UpdateNodesBasedOnSyncState(nodesWithSyncStatus []*data.NodeData) {
+	syncedNodes, outOfSyncNodes := computeSyncedAndOutOfSyncNodes(nodesWithSyncStatus)
+
+	bnp.mutNodes.Lock()
+	defer bnp.mutNodes.Unlock()
+
+	for _, outOfSyncNode := range outOfSyncNodes {
+		if len(bnp.nodes[outOfSyncNode.ShardId]) < 2 {
+			log.Warn("cannot remove observer as not enough will remain in shard",
+				"address", outOfSyncNode.Address,
+				"shard", outOfSyncNode.ShardId)
+			continue
+		}
+
+		bnp.removeNodeUnprotected(outOfSyncNode)
+	}
+
+	bnp.addSyncedNodesUnprotected(syncedNodes)
+}
+
+func computeSyncedAndOutOfSyncNodes(nodes []*data.NodeData) ([]*data.NodeData, []*data.NodeData) {
+	syncedNodes := make([]*data.NodeData, 0)
+	outOfSyncNodes := make([]*data.NodeData, 0)
+	for _, node := range nodes {
+		if node.IsSynced {
+			syncedNodes = append(syncedNodes, node)
+			continue
+		}
+
+		log.Warn("observer is out of sync", "address", node.Address, "shard", node.ShardId)
+		outOfSyncNodes = append(outOfSyncNodes, node)
+	}
+
+	return syncedNodes, outOfSyncNodes
+}
+
+func (bnp *baseNodeProvider) addSyncedNodesUnprotected(receivedSyncedNodes []*data.NodeData) {
+	for _, node := range receivedSyncedNodes {
+		if bnp.isReceivedSyncedNodeExistent(node) {
+			continue
+		}
+
+		bnp.allNodes = append(bnp.allNodes, node)
+	}
+
+	bnp.nodes = nodesSliceToShardedMap(bnp.allNodes)
+}
+
+func (bnp *baseNodeProvider) isReceivedSyncedNodeExistent(receivedNode *data.NodeData) bool {
+	for _, node := range bnp.allNodes {
+		if node.Address == receivedNode.Address && node.ShardId == receivedNode.ShardId {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (bnp *baseNodeProvider) addToOutOfSyncUnprotected(node *data.NodeData) {
+	if bnp.outOfSyncNodes == nil {
+		bnp.outOfSyncNodes = make([]*data.NodeData, 0)
+	}
+
+	for _, oosNode := range bnp.outOfSyncNodes {
+		if oosNode.Address == node.Address && oosNode.ShardId == node.ShardId {
+			log.Warn("[programming error] -> node is already in out of sync list", "address", node.Address, "shard ID", node.ShardId)
+			return
+		}
+	}
+
+	bnp.outOfSyncNodes = append(bnp.outOfSyncNodes, node)
+}
+
+func (bnp *baseNodeProvider) removeNodeUnprotected(node *data.NodeData) {
+	bnp.removeNodeFromShardedMapUnprotected(node)
+	bnp.removeNodeFromAllNodesUnprotected(node)
+	bnp.addToOutOfSyncUnprotected(node)
+}
+
+func (bnp *baseNodeProvider) removeNodeFromShardedMapUnprotected(node *data.NodeData) {
```

Comment:
not necessarily in this PR, but i think we can refactor these 2 functions to remove duplicated parts

Reply:
I'm not sure if the common code can be extracted